### DISCUSSION TITLE
Download missing beatmap when importing scores

### DISCUSSION THREAD

**Discussion by PercyDan54** [ppy/osu#14029] [Q&A] [answered]
I'm currently working on this, and managed to get this
https://user-images.githubusercontent.com/50285552/127169378-343ff690-3ec5-4589-8734-62dd561eecf2.mp4

But making these two notifications makes no sense, and I'm lacking UX design ideas.
Ideas:
1. Combine these two notifications into one. But there's a problem:
I have to wait for the beatmap query to complete then show the notification.

2. Implement a screen for displaying and downloading missing beatmaps similar to the one in `SoloSpectator`

**Comment by peppy** [comment, marked as answer]
What is the intention here? A flow where a user is importing one single score to immediately watch it? If more than one scores are imported does this flow also need to exist?

**Reply by PercyDan54** [reply]
You can drag multiple scores to import at once, and notifications would be spamming

**Reply by peppy** [reply]
So this flow you are planning to add is just for single score imports, correct? Which should be considered a special case where the user wants to watch the score immediately?

**Reply by PercyDan54** [reply]
If we're going for the notification then yes

**Reply by peppy** [reply]
I think the screen flow similar to spectator is the best method. There needs to be some kind of loading screen while the beatmap is obtained, and that suits the use case for now.

**Reply by PercyDan54** [reply]
There's a problem that only one score at a time will be passed to `ScoreManager.CreateModel()` a time, so idk how to handle multiple imports.

Actually I also prefer the screen flow, the notification is just the basic implemention.

**Reply by PercyDan54** [reply]
Is this fine? Most code is just copied from the spectator screen, maybe needs to split it out to avoid duplicate codes

https://user-images.githubusercontent.com/50285552/127169378-343ff690-3ec5-4589-8734-62dd561eecf2.mp4

**Reply by bdach** [reply]
There are still two notifications shown, and I'm not sure what "replay mode" is supposed to mean.

I think ideally the splash screen would contain the info about the score that the user is attempted to watch (so username, total score, etc.)

**Reply by PercyDan54** [reply]
Like this?
The score is not imported yet, I just forced `LegacyScoreDecoder` to decode the score ignoring `BeatmapNotFoundException`

Maybe change the "replay mode" text here to "beatmap info"?

![image](https://user-images.githubusercontent.com/50285552/127264922-4e912423-9153-4a0d-b278-74b8501230bd.png)

**Reply by peppy** [reply]
Seems okay for now.

**Comment by frenzibyte** [comment]
PR opened: https://github.com/ppy/osu/pull/14044

**Comment by monochrome22** [comment]
This UX flow looks convoluted to me. In my opinion, it should be as simple as clicking the notification and getting the beatmap.

This is how I imagine it:
1
![Scoreimport1](https://github.com/ppy/osu/assets/122969285/99bec2b4-0608-474a-af7f-8132ea32143b)
2
![scoreimport2](https://github.com/ppy/osu/assets/122969285/b5dc2741-8919-44da-8b7e-6a0d06adcfe8)
3
![Scoreimport3](https://github.com/ppy/osu/assets/122969285/27809510-1378-44eb-85a2-15eca0947534)

Each notification should disappear when the next one appears.

---

[In the case of the beatmap not existing](https://github.com/ppy/osu/pull/14044#issuecomment-888403217), it should instead write that the beatmap can't be downloaded:

![Scoreimport4](https://github.com/ppy/osu/assets/122969285/067d0ed3-c077-4a3c-b583-e9ed671798db)

If the user is not logged in or doesn't have an internet connection, then it should display like this:
1
![Scoreimport6](https://github.com/ppy/osu/assets/122969285/9679683d-445b-4a26-bbe9-32a7c43c5ebd)
After logging in, if the user has not dismissed the first notification before logging in, then it will display this:
2
![Scoreimport7](https://github.com/ppy/osu/assets/122969285/a77cdcb7-ea42-4a6f-ada9-c5d9cc03da16)

(I don't know if downloading is possible while not logged in. Beatmap listing requires you to be logged in.)

---

>1.Combine these two notifications into one. But there's a problem:
>I have to wait for the beatmap query to complete then show the notification.

If this is a problem, then have the importing notification load for a few seconds if necessary, instead of immediately displaying it:

![Scoreimport5](https://github.com/ppy/osu/assets/122969285/48d2691e-9af5-4ffa-9311-e480f3f3f2a0)

---

This flow is more simple and it will allow the user to easily be able to do others things while waiting for the map to download. This is especially good for people with slower internets; they can play a map or watch another replay while waiting for the score map to download.
Having another screen makes sense for spectator, because you're not going to be doing something else while sitting in a multiplayer lobby, but it's not suitable for score imports.

**Comment by nfnfgo** [comment]
Also think the current error message when replay importing failed is not clear enough, and **this vague error message could cause issue and lots of frustration for a relatively new user who doesn't know they need to download the beatmap in advance**, because this error message doesn't provide any useful info on how to solve this error.

![Screenshot of error message in osu lazer](https://github.com/user-attachments/assets/42ff269b-55e1-4bde-bff2-988e699d9c04)
(Error message in newest lazer client `2024.1219.2`)

Even if not considering auto download the map, the error message itself should be more informative, and provide the user with reason (and possible solution if exists) of the import failure:

- Replay file itself is invalid?
- Corresponding beatmap not downloaded? (In this case, at least a link to the missing beatmap should be provided)
- Some other reason?

and not just tell user to check out the log file. 

------

Personally, I do think "click this notification to download" method stated by @monochrome22 a good and intuitive way to address this issue. Moreover, I think it **also make sense to directly start the download** without waiting user to click the notification **if user toggled on "Download missing map automatically" in settings** (check the screenshot below).

![Settings screenshot](https://github.com/user-attachments/assets/bb9f7168-5076-4360-8f83-e2520dd88e8d)

For bulk importing, there could be two strategies:

- "Download missing map automatically" ON: Directly download ALL missing beatmap.
- "Download missing map automatically" OFF: First import all scores that has corresponding beatmap downloaded, then allow users to click the notification to download all missing beatmap then import the rest of the scores if they want.
